### PR TITLE
CopilotChat: Add Qdrant memory store and File system chat store

### PR DESCRIPTION
**Pull request by adrianwyatt**
### Motivation and Context
- Need to support Qdrant as a memory store in CopilotChat to showcase vector DBs
- Wanted to added a lightweight persistent chat store that can run locally.

### Description
- Added Qdrant memory store support in CopilotChat
- Added FileSystem chat store in CopilotChat
- Made swapping memory stores and chat stores available in configuration
- Removed "*Config" naming paradigm in CopilotChat appsettings (i.e. "EmbeddingConfig" -> "Embedding")

### CODE REVIEW

File: samples/apps/copilot-chat-app/webapi/Program.cs [hunk 1]
```diff
@@ -139,26 +160,42 @@ private static void AddSemanticKernelServices(this IServiceCollection services,
...
+                    new FileInfo(Path.Combine(directory, $"{Path.GetFileNameWithoutExtension(fullPath)}_sessions{Path.GetExtension(fullPath)}")));
+                chatMessageInMemoryContext = new FileSystemContext<ChatMessage>(
+                    new FileInfo(Path.Combine(directory, $"{Path.GetFileNameWithoutExtension(fullPath)}_messages{Path.GetExtension(fullPath)}")));
+                break;
+
+            case "COSMOS":
+                CosmosConfig cosmosConfig = configuration.GetSection("ChatStore:Cosmos").Get<CosmosConfig>();
+#pragma warning disable CA2000 // Dispose objects before losing scope - objects are singletons for the duration of the process and disposed when the process exits.
+                chatSessionInMemoryContext = new CosmosDbContext<ChatSession>(
+                    cosmosConfig.ConnectionString, cosmosConfig.Database, "chat_sessions");
```
Comment (TaoChenOSU) [inline samples/apps/copilot-chat-app/webapi/Program.cs:188]: Do we want to hard code the container name here? Could we move it to the config as well?

File: samples/apps/copilot-chat-app/webapi/Program.cs [hunk 2]
```diff
@@ -139,26 +160,42 @@ private static void AddSemanticKernelServices(this IServiceCollection services,
...
+                chatMessageInMemoryContext = new CosmosDbContext<ChatMessage>(
+                    cosmosConfig.ConnectionString, cosmosConfig.Database, "chat_messages");
```
Comment (TaoChenOSU) [inline samples/apps/copilot-chat-app/webapi/Program.cs:190]: Same as above

File: samples/apps/copilot-chat-app/webapi/Storage/FileSystemContext.cs [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        lock (this._fileStorageLock)
+        {
+            if (!fileInfo.Exists)
+            {
+                fileInfo.Directory!.Create();
+                File.WriteAllText(fileInfo.FullName, "{}");
+            }
+
+            using FileStream fileStream = File.Open(
+            path: fileInfo.FullName,
```
Comment (TaoChenOSU) [inline samples/apps/copilot-chat-app/webapi/Storage/FileSystemContext.cs:126]: nit: tab

File: samples/apps/copilot-chat-app/webapi/Program.cs (hunk 1 above)
Comment (craigomatic) [reply samples/apps/copilot-chat-app/webapi/Program.cs:188]: At a minimum we would need to switch the hard coded name to one that does not have an underscore, see https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/how-to-dotnet-create-container for allowed chars in name.

File: samples/apps/copilot-chat-app/webapi/Program.cs [hunk 4]
```diff
@@ -95,11 +96,31 @@ private static void AddServices(IServiceCollection services, ConfigurationManage
...
+        services.AddScoped<Kernel>();
+
+        // Add a semantic memory store only if we have a valid embedding config
+        AIServiceConfig embeddingConfig = configuration.GetSection("Embedding").Get<AIServiceConfig>();
         if (embeddingConfig?.IsValid() == true)
         {
-            services.AddSingleton<IMemoryStore, VolatileMemoryStore>();
+            switch (configuration["MemoriesStore:Type"].ToUpperInvariant())
+            {
+                case "VOLATILE":
```
Comment (craigomatic) [inline samples/apps/copilot-chat-app/webapi/Program.cs:108]: Does it make sense to pull these out into constants?

File: samples/apps/copilot-chat-app/webapi/Program.cs (hunk 1 above)
Comment (adrianwyatt) [reply samples/apps/copilot-chat-app/webapi/Program.cs:188]: Moved to config and removed underscore.

File: samples/apps/copilot-chat-app/webapi/Program.cs (hunk 2 above)
Comment (adrianwyatt) [reply samples/apps/copilot-chat-app/webapi/Program.cs:190]: Moved to config and removed underscore.

File: samples/apps/copilot-chat-app/webapi/appsettings.json [hunk 5]
```diff
@@ -1,32 +1,49 @@
...
+    /* "Key": // dotnet user-secrets set "Embedding:Key" "MY_EMBEDDING_KEY" */
+  },
+
+  "ChatStore": {
+    "Type": "volatile", // Which type of chat store to use - supported values are "volatile", "filesystem", or "cosmos".
+    "Filesystem": {
+      "FilePath": "./data/chatstore.json"
+    },
+    "Cosmos": {
+      "Database": "adribona-CopilotChat"
```
[the remark this answers is not in the document]
Comment (adrianwyatt) [reply samples/apps/copilot-chat-app/webapi/appsettings.json:34]: Nope! Updated.

**Review by TaoChenOSU — APPROVED** [review]

**Pull request merged**